deployment/ccip/changeset/cs_chain_contracts.go

```diff
@@ -351,6 +353,283 @@ func UpdateOnRampsDestsChangeset(e deployment.Environment, cfg UpdateOnRampDests
 	}}, nil
 }
 
+type UpdateOnRampDynamicConfig struct {
+	// UpdatesByChain is a mapping of source -> update.
+	UpdatesByChain map[uint64]OnRampDynamicConfigUpdate
+	// Disallow mixing MCMS/non-MCMS per chain for simplicity.
+	// (can still be achieved by calling this function multiple times)
+	MCMS *MCMSConfig
+}
+
+type OnRampDynamicConfigUpdate struct {
+	FeeQuoter          common.Address
+	MessageInterceptor common.Address
+	FeeAggregator      common.Address
+	AllowlistAdmin     common.Address
+}
+
+func (cfg UpdateOnRampDynamicConfig) Validate() error {
+	for chainSel, config := range cfg.UpdatesByChain {
+		if config.FeeQuoter == (common.Address{}) {
+			return fmt.Errorf("FeeQuoter is not specified for chain=%d", chainSel)
+		}
+		if config.FeeAggregator == (common.Address{}) {
+			return fmt.Errorf("FeeAggregator is not specified for chain=%d", chainSel)
+		}
+	}
+	return nil
+}
+
+func UpdateOnRampDynamicConfigChangeset(e deployment.Environment, cfg UpdateOnRampDynamicConfig) (deployment.ChangesetOutput, error) {
+	if err := cfg.Validate(); err != nil {
+		return deployment.ChangesetOutput{}, err
+	}
+	onchain, err := LoadOnchainState(e)
+	if err != nil {
+		return deployment.ChangesetOutput{}, err
+	}
+	var batches []timelock.BatchChainOperation
+	timelocks := make(map[uint64]common.Address)
+	proposers := make(map[uint64]*gethwrappers.ManyChainMultiSig)
+	ctx := e.GetContext()
+	for chainSel, update := range cfg.UpdatesByChain {
+		txOps := e.Chains[chainSel].DeployerKey
+		//nolint:fatcontext // context is stored outside the loop
+		txOps.Context = ctx
+		if cfg.MCMS != nil {
+			txOps = deployment.SimTransactOpts()
+		}
+		onRamp := onchain.Chains[chainSel].OnRamp
+		tx, err := onRamp.SetDynamicConfig(txOps, onramp.OnRampDynamicConfig{
+			FeeQuoter:              update.FeeQuoter,
+			ReentrancyGuardEntered: false,
+			MessageInterceptor:     update.MessageInterceptor,
+			FeeAggregator:          update.FeeAggregator,
+			AllowlistAdmin:         update.AllowlistAdmin,
+		})
+		if err != nil {
+			return deployment.ChangesetOutput{}, err
+		}
+		if cfg.MCMS == nil {
+			if _, err := deployment.ConfirmIfNoError(e.Chains[chainSel], tx, err); err != nil {
+				return deployment.ChangesetOutput{}, err
```

Comment:
Need to use [DecodedErrFromABIIfDataErr](https://github.com/smartcontractkit/chainlink/blob/develop/deployment/environment.go#L206) for all write txs for better error reporting